rclc/include/rclc/executor.h

```diff
@@ -0,0 +1,387 @@
+  rclc_executor_handle_counters_t info;
+  /// timeout in nanoseconds for rcl_wait() used in rclc_executor_spin_once(). Default 100ms
+  uint64_t timeout_ns;
+  /// timepoint used for spin_period()
+  rcutils_time_point_value_t invocation_time;
+  /// trigger function, when to process new data
+  rclc_executor_trigger_t trigger_function;
+  /// application specific data structure for trigger function
+  void * trigger_object;
+  /// data communication semantics
+  rclc_executor_semantics_t data_comm_semantics;
+} rclc_executor_t;
+
+/**
+ *  Return a rclc_executor_t struct with pointer members initialized to `NULL`
+ *  and member variables to 0.
+ */
+rclc_executor_t
+rclc_executor_get_zero_initialized_executor(void);
+/**
+ *  Initializes an executor.
+ *  It creates a dynamic array with size \p number_of_handles using the
+ *  \p allocator.
+ *
+ *
+ *  * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | Yes
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param[inout] e preallocated rclc_executor_t
+ * \param[in] context RCL context
+ * \param[in] number_of_handles size of the handle array
+ * \param[in] allocator allocator for allocating memory
+ * \return `RCL_RET_OK` if the executor was initialized successfully
+ * \return `RCL_RET_INVALID_ARGUMENT` if any null pointer as argument
+ * \return `RCL_RET_ERROR` in case of failure
+ */
+rcl_ret_t
+rclc_executor_init(
+  rclc_executor_t * executor,
+  rcl_context_t * context,
+  const size_t number_of_handles,
+  const rcl_allocator_t * allocator);
+
+/**
+ *  Set timeout in nanoseconds for rcl_wait (called during {@link rclc_executor_spin_once()}).
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | No
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param [inout] executor pointer to an initialized executor
+ * \param [in] timeout_ns  timeout in nanoseconds for the rcl_wait (DDS middleware)
+ * \return `RCL_RET_OK` if timeout was set successfully
+ * \return `RCL_RET_INVALID_ARGUMENT` if \p executor is a null pointer
+ * \return `RCL_RET_ERROR` in an error occured
+ */
+rcl_ret_t
+rclc_executor_set_timeout(
+  rclc_executor_t * executor,
+  const uint64_t timeout_ns);
+
+/**
+ *  Set data communication semantics
+ *
+ * <hr>
+ * Attribute          | Adherence
+ * ------------------ | -------------
+ * Allocates Memory   | No
+ * Thread-Safe        | No
+ * Uses Atomics       | No
+ * Lock-Free          | Yes
+ *
+ * \param [inout] executor pointer to an initialized executor
+ * \param [in] valid semantics value as defined in enum type {@link rclc_executor_semantics_t}
+ * \return `RCL_RET_OK` if semantics was set successfully
+ * \return `RCL_RET_INVALID_ARGUMENT` if \p executor is a null pointer
+ */
+
```

Comment:
```suggestion
```

Reply:
fixed in 54f0bbc54a33cc601764ab1d53fe3119d9e61c8b